msal/src/main/java/com/microsoft/identity/client/SingleAccountPublicClientApplication.java

```diff
@@ -66,66 +107,270 @@ private void initializeSharedPreferenceFileManager(@NonNull final Context contex
+                null, // loginHint
+                null // claimsRequest
+        );
+    }
+
     @Override
-    public void removeCurrentAccount(@NonNull final TaskCompletedCallbackWithError<Boolean, Exception> callback) {
+    protected ILocalAuthenticationCallback getLocalAuthenticationCallback(final AuthenticationCallback authenticationCallback) {
+
+        return new ILocalAuthenticationCallback() {
+
+            @Override
+            public void onSuccess(ILocalAuthenticationResult localAuthenticationResult) {
+
+                //Get Local Authentication Result then check if the current account is set or not
+                MultiTenantAccount newAccount =  getAccountFromICacheRecordList(localAuthenticationResult.getCacheRecordWithTenantProfileData());
+
+                if(didExistingCurrentAccountChange(newAccount)){
+                    //Throw on Error with UserMismatchException
+                    authenticationCallback.onError(new MsalClientException(MsalClientException.CURRENT_ACCOUNT_MISMATCH));
+                    return;
+                }else{
+                    persistCurrentAccount(localAuthenticationResult.getCacheRecordWithTenantProfileData());
+                }
+
+                IAuthenticationResult authenticationResult = AuthenticationResultAdapter.adapt(localAuthenticationResult);
+                authenticationCallback.onSuccess(authenticationResult);
+            }
+
+            @Override
+            public void onError(BaseException exception) {
+                MsalException msalException = MsalExceptionAdapter.msalExceptionFromBaseException(exception);
+                authenticationCallback.onError(msalException);
+            }
+
+            @Override
+            public void onCancel() {
+                authenticationCallback.onCancel();
+            }
+        };
+    }
+
+    private boolean didExistingCurrentAccountChange(MultiTenantAccount newAccount){
+        MultiTenantAccount persistedAccount = getPersistedCurrentAccount();
+        return (persistedAccount != null || newAccount.getId() != persistedAccount.getId()) ? true : false;
+
+    }
+
+    @Override
+    public void signOut(@NonNull final TaskCompletedCallbackWithError<Boolean, Exception> callback){
+        final String methodName = ":signOut";
         final PublicClientApplicationConfiguration configuration = getConfiguration();
 
-        new BrokerMsalController().removeAccountFromSharedDevice(
-                configuration,
-                new TaskCompletedCallbackWithError<Boolean, Exception>() {
-                    @Override
-                    public void onTaskCompleted(Boolean success) {
-                        if (success) {
-                            persistCurrentAccount(null);
+
+        try{
+            if (mIsSharedDevice) {
+                removeAccountFromSharedDevice(callback, configuration);
+                return;
+            }
+
+
+            ApiDispatcher.initializeDiagnosticContext();
```

Comment:
Need to remove this call. All the initializeDiagnosticContext() is now called before the commands excution at inside ApiDispatcher. Duplicated initializeDiagnosticContext() calls will result multiple correlationId for one request.

Reply:
I assume the correlationId initialization should either happen at the PCA level or the ApiDispatcher level.

Reply:
Makes sense to me.  I copied this code from multiaccountpca... so we likely need to update there as well.

Reply:
removed.